File: kotlin-node/karakum/injections/WritableMembersInjection.js
import ts from "typescript";
import * as karakum from "karakum";

export default {
    setup(context) {
        this.readableStreamMemberNodes = new Map()
    },

    traverse(node) {
        const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"

        if (
            sourceFileName.endsWith("globals.d.ts")
            && ts.isMethodSignature(node)
            && ts.isIdentifier(node.name)
            && (
                node.name.text === "write"
                || (
                    node.name.text === "end"
                    && node.parameters.length > 1
                )
            )

            && ts.isInterfaceDeclaration(node.parent)
            && node.parent.name.text === "WritableStream"
        ) {
            this.readableStreamMemberNodes.set(node.name.text + node.parameters.length, node)
        }

        if (
            sourceFileName.endsWith("globals.d.ts")
            && ts.isPropertySignature(node)
            && ts.isIdentifier(node.name)
            && node.name.text === "writable"

            && ts.isInterfaceDeclaration(node.parent)
            && node.parent.name.text === "WritableStream"
        ) {
            this.readableStreamMemberNodes.set(node.name.text, node)
        }
    },

    render(node, context, next) {
        return null
    },

    inject(node, context, render) {
        const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"

        if (
            sourceFileName.endsWith("stream.d.ts")
            && ts.isClassDeclaration(node)
            && node?.name.text === "Writable"
            && !context.static
        ) {
            return Array.from(this.readableStreamMemberNodes.values())
                .map(member => {
                    if (ts.isPropertySignature(member)) {
                        return `override ${render(member)}`
                    }

                    const name = karakum.escapeIdentifier(render(member.name))

                    const typeParameters = member.typeParameters
                        ?.map(typeParameter => render(typeParameter))
                        ?.join(", ")

                    const returnType = member.type && render(member.type)

                    return karakum.convertParameterDeclarations(member, context, render, {
                        strategy: "function",
                        defaultValue: "",
                        template: parameters => {
                            return `override fun ${karakum.ifPresent(typeParameters, it => `<${it}> `)}${name}(${parameters})${karakum.ifPresent(returnType, it => `: ${it}`)}`
                        }
                    })
                })
                .join("\n")
        }

        return null
    },

    generate(context) {
        return []
    }
}

File: kotlin-node/karakum/src/jsMain/kotlin/wrappersgenerator/node/injections/WritableMembersInjection.kt
package wrappersgenerator.node.injections

import arrow.core.raise.impure
import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.*
import io.github.sgrishchenko.karakum.extension.plugins.ParameterDeclarationStrategy
import io.github.sgrishchenko.karakum.extension.plugins.convertParameterDeclarations
import io.github.sgrishchenko.karakum.extension.plugins.function
import io.github.sgrishchenko.karakum.util.getParentOrNull
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import typescript.*

class WritableMembersInjection : Injection {
    private val writableStreamMemberNodes = mutableListOf<Node>()

    override suspend fun setup(context: Context) = Unit

    override suspend fun traverse(node: Node, context: Context) = impure {
        val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName
        ensure(sourceFileName.endsWith("globals.d.ts"))

        nullable {
            ensure(isMethodSignature(node))

            val name = ensureNotNull(node.name)
            ensure(isIdentifier(name))
            ensure(
                name.text == "write"
                        || (name.text == "end" && node.parameters.asArray().size > 1)
            )

            val interfaceNode = ensureNotNull(node.getParentOrNull())
            ensure(isInterfaceDeclaration(interfaceNode))
            ensure(interfaceNode.name.text == "WritableStream")

            writableStreamMemberNodes += node
        } ?: nullable {
            ensure(isPropertySignature(node))

            val name = ensureNotNull(node.name)
            ensure(isIdentifier(name))
            ensure(name.text == "writable")

            val interfaceNode = ensureNotNull(node.getParentOrNull())
            ensure(isInterfaceDeclaration(interfaceNode))
            ensure(interfaceNode.name.text == "WritableStream")

            writableStreamMemberNodes += node
        }

    }

    override suspend fun render(node: Node, context: Context, next: Render<Node>) = null

    override suspend fun inject(node: Node, context: InjectionContext, render: Render<Node>) = nullable {
        ensure(context.type == InjectionType.MEMBER)

        val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName
        ensure(sourceFileName.endsWith("stream.d.ts"))

        ensure(isClassDeclaration(node))
        ensure(node.name?.text == "Writable")

        writableStreamMemberNodes
            .mapNotNull { member ->
                nullable {
                    ensure(isPropertySignature(member))

                    "override ${render(member)}"
                } ?: nullable {
                    ensure(isMethodSignature(member))

                    val name = render(member.name)

                    val typeParameters = member.typeParameters?.asArray()
                        ?.map { render(it) }
                        ?.joinToString(", ")

                    val returnType = member.type?.let { render(it) }

                    convertParameterDeclarations(
                        member, context, render,
                        ParameterDeclarationStrategy.function,
                        defaultValue = "",
                    ) { parameters, _ ->
                        "override fun ${ifPresent(typeParameters) { "<${it}> " }}${name}(${parameters})${ifPresent(returnType) { ": $it" }}"
                    }
                }
            }
            .toTypedArray()
    }

    override suspend fun generate(context: Context, render: Render<Node>) = emptyArray<GeneratedFile>()
}
